experimental/xformer/Transforms/ReplaceFCWithConv2D.cpp

```
@@ -0,0 +1,203 @@
+              .getElementType()
+              .isa<quant::QuantizedType>() &&
+          fcOp.input()
+              .getType()
+              .cast<ShapedType>()
+              .getElementType()
+              .cast<quant::QuantizedType>()
+              .isSigned() &&
+          fcOp.input()
+                  .getType()
+                  .cast<ShapedType>()
+                  .getElementType()
+                  .cast<quant::QuantizedType>()
+                  .getStorageTypeIntegralWidth() == 8)) {
+      return failure();
+    }
+
+    // Filter type must be QI8
+    if (!(fcOp.filter()
+              .getType()
+              .cast<ShapedType>()
+              .getElementType()
+              .isa<quant::QuantizedType>() &&
+          fcOp.filter()
+              .getType()
+              .cast<ShapedType>()
+              .getElementType()
+              .cast<quant::QuantizedType>()
+              .isSigned() &&
+          fcOp.filter()
+                  .getType()
+                  .cast<ShapedType>()
+                  .getElementType()
+                  .cast<quant::QuantizedType>()
+                  .getStorageTypeIntegralWidth() == 8)) {
+      return failure();
+    }
+
+    // TODO: What to do if no bias?
+    // Bias type must be QI32
+    if (!(fcOp.bias()
+              .getType()
+              .cast<ShapedType>()
+              .getElementType()
+              .isa<quant::QuantizedType>() &&
+          fcOp.bias()
+              .getType()
+              .cast<ShapedType>()
+              .getElementType()
+              .cast<quant::QuantizedType>()
+              .isSigned() &&
+          fcOp.bias()
+                  .getType()
+                  .cast<ShapedType>()
+                  .getElementType()
+                  .cast<quant::QuantizedType>()
+                  .getStorageTypeIntegralWidth() == 32)) {
+      return failure();
+    }
+
+    if (fcOp.weights_format() != "DEFAULT") {
+      return failure();
```

could we return an error via  fcOp.emitOpError() << "stuff"?

---

for all failures

---

I think you are right, this is a pattern match fail which isn't error

---

maybe

  /// Emit a remark about this operation, reporting up to any diagnostic
  /// handlers that may be listening.
  InFlightDiagnostic emitRemark(const Twine &message = {});


Would be a good option